### PR TITLE
Deprecate 'generate openapi', add 'generate crds'

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
This PR deprecates the  `operator-sdk generate openapi` command and adds `operator-sdk generate crds`

**Motivation for the change:**
The `zz_generated.openapi.go` files are rarely required, so the SDK should be less opinionated about creating them for users.

This change also aligns the SDK more closely with `kubebuilder`, which does not generate `zz_generated.openapi.go` by default.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,10 +12,12 @@
 - Updated `pkg/test/e2eutil.WaitForDeployment()` and `pkg/test/e2eutil.WaitForOperatorDeployment()` to successfully complete waiting when the available replica count is _at least_ (rather than exactly) the minimum replica count required. ([#2248](https://github.com/operator-framework/operator-sdk/pull/2248))
 - Replace in the Ansible based operators module tests `k8s_info` for `k8s_facts` which is deprecated. ([#2168](https://github.com/operator-framework/operator-sdk/issues/2168))
 - Upgrade the Ansible version from `2.8` to `2.9` on the Ansible based operators image. ([#2168](https://github.com/operator-framework/operator-sdk/issues/2168))
+- **Breaking change:** Renamed `operator-sdk generate openapi` to `operator-sdk generate crds`. ([#2276](https://github.com/operator-framework/operator-sdk/pull/2276))
```
Comment (camilamacedo86) [inline CHANGELOG.md]: I am not sure if we should move forward with. See that kubebuilder is not supporting it and users are asking for [here](https://github.com/kubernetes-sigs/kubebuilder/issues/1129#issuecomment-558704012) and a track issue from them [here](https://github.com/kubernetes-sigs/kubebuilder/issues/1081) which in my understand shows that require it as well.

Comment (estroz) [reply CHANGELOG.md]: The first issue is fair; based on feedback in #2042 some users do not use `zz_generated.openapi.go`, which, on top of it not being necessary for API code unlike `zz_generated.deepcopy.go`, is why I decided to push for removing it. We should either:

1. Add documentation on building and using [`openapi-gen`](https://github.com/kubernetes/kube-openapi/tree/master/cmd/openapi-gen) to generate the files yourself, which is fairly straightforward.
2. Keep `generate openapi` as a command but move CRD generation into `generate crd`.

I vote for option 1 since we can offer Go OpenAPI code by wrapping `openapi-gen` in a Makefile recipe (in the future).

The second issue describes something we do not currently support in CRDs anyway since we use the same tool as kubebuilder to generate CRDs, [`controller-gen`](https://github.com/kubernetes-sigs/controller-tools/blob/6eef398/cmd/controller-gen/main.go) from controller-tools. Generating `zz_generated.openapi.go` does not change the state of supported annotations for CRD manifest validation in the SDK.

We agreed to make this change because it de-couples CRD generation from OpenAPI spec generation, two different albeit related processes. This change will help users debug errors more easily as well.

File: hack/generate/gen-cli-doc.go [hunk 2]
```diff
@@ -32,7 +33,19 @@ func main() {
 		log.Fatalf("Failed to get current directory: %v", err)
 	}
 
-	err = doc.GenMarkdownTree(root, currentDir+"/doc/cli")
+	docPath := filepath.Join(currentDir, "doc", "cli")
```
Comment (estroz) [inline hack/generate/gen-cli-doc.go]: @joelanford can you make a separate PR with this change?

Comment (joelanford) [reply hack/generate/gen-cli-doc.go]: See https://github.com/operator-framework/operator-sdk/pull/2278. I'll remove this change from this PR.

File: CHANGELOG.md (hunk 1 above)
Comment (joelanford) [reply CHANGELOG.md]: Another option would be to add `generate crds` but keep `generate openapi` as is and deprecate it. We could hide it from the help output and log a deprecation message when it is called.

Then in a later version it could be removed.

Comment (joelanford) [reply CHANGELOG.md]: And we could/should still do one of the options @estroz suggested.

Comment (estroz) [reply CHANGELOG.md]: @joelanford I like the combination of what you suggested and my first option. The docs referencing Go OpenAPI spec generation (I believe the only documentation doing so is the one being changed in this PR already) should warn that the subcommand is deprecated and users should switch to installing and running `openapi-gen` themselves if they still want that spec file.

Comment (shawn-hurley) [reply CHANGELOG.md]: I think the idea of phasing out the old command with a deprecation warning. 

@estroz @joelanford we need to make it easy to download and use the openapi-gen command if that is the guidance, we either do it for the user and attempt to install the binary into a known location and give them the option to supply their own openapi-gen. 

Could we vendor the openapi-gen command and use it in the operator-sdk command? was that an option already looked?

Comment (estroz) [reply CHANGELOG.md]: @shawn-hurley we currently vendor the exposed `openapi-gen` generator in `generate openapi`. If we want to install this binary for the user, I suggest this as the first candidate for a Makefile:
```make
openapi-gen:
	which ./bin/openapi-gen > /dev/null || go build -o ./bin/openapi-gen k8s.io/kube-openapi/cmd/openapi-gen
	./bin/openapi-gen ...
```
Otherwise documentation to the effect of the above script should suffice. I'd rather have documentation than a make recipe.

Comment (joelanford) [reply CHANGELOG.md]: By the way, our vendoring of the code generation libraries is what causes the GOROOT and panic issues described in this issue: https://github.com/operator-framework/operator-sdk/issues/1854#issuecomment-525132306

If we direct users to build `openapi-gen` from source like Eric suggests, we'll help them avoid that issue, at least for the openapi generation (I think it will still be an issue for deepcopy generation).

Comment (joelanford) [reply CHANGELOG.md]: Alright, updated the PR to keep `generate openapi`, but to hide it from help and deprecate it.

File: doc/sdk-cli-reference.md [hunk 3]
```diff
@@ -134,6 +134,21 @@ pkg/apis/app/v1alpha1/
 
 ### openapi
```
Comment (estroz) [inline doc/sdk-cli-reference.md]: Side note: should we remove this doc entirely now that we auto-generate CLI docs under `doc/cli`?

File: hack/generate/gen-cli-doc.go [hunk 4]
```diff
@@ -33,19 +32,7 @@ func main() {
...
-	// are removed.
-	if err := os.RemoveAll(docPath); err != nil {
-		log.Fatalf("Failed to remove existing generated docs: %v", err)
-	}
-	if err := os.MkdirAll(docPath, 0755); err != nil {
-		log.Fatalf("Failed to re-create docs directory: %v", err)
-	}
-
-	err = doc.GenMarkdownTree(root, docPath)
+	err = doc.GenMarkdownTree(root, currentDir+"/doc/cli")
```
Comment (estroz) [inline hack/generate/gen-cli-doc.go]: This reverts #2278.

File: doc/sdk-cli-reference.md (hunk 3 above)
Comment (joelanford) [reply doc/sdk-cli-reference.md]: Yeah, probably.

@theishshah is there anything else we need to do before we remove it?

File: hack/generate/gen-cli-doc.go (hunk 4 above)
Comment (joelanford) [reply hack/generate/gen-cli-doc.go]: Huh.. clearly I messed that up.

Comment (joelanford) [reply hack/generate/gen-cli-doc.go]: Ok fixed that. Good catch.

File: cmd/operator-sdk/generate/crds.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+package generate
+
+import (
+	"fmt"
+
+	"github.com/operator-framework/operator-sdk/cmd/operator-sdk/internal/genutil"
+	"github.com/spf13/cobra"
+)
+
+func newGenerateCRDsCmd() *cobra.Command {
```
Comment (shawn-hurley) [inline cmd/operator-sdk/generate/crds.go]: can we add a flag for passing in the location of the CRD's?

Comment (joelanford) [reply cmd/operator-sdk/generate/crds.go]: Once this merges, I'll make an issue to discuss this idea for a follow-up.

File: doc/sdk-cli-reference.md (hunk 3 above)
Comment (camilamacedo86) [reply doc/sdk-cli-reference.md]: If we are just deprecated the command then: 
IHMO:
- We should not remove/change the docs
- We just need to add on top of the func used by it which will be removed the comments that it is deprecated for have the func strikethrough as:

Screenshot 2019-12-04 at 21 39 44

Comment: `// Deprecated: xpto`

- Also, we need to provide a doc with the info for users knowhow them can customize the projects to still be able to gen and use the open API.

Then, should not be better have a PR to deprecated and one to add the new command?

Comment (estroz) [reply doc/sdk-cli-reference.md]: We want existing users who have written scripts around `generate openapi` to still be able to use the command while they migrate to `generate crds`, but new users to use `generate crds`. By removing docs we can ensure visibility of `generate crds` without confusing anyone.

The "documentation" explaining deprecation and new behavior is [here](https://github.com/operator-framework/operator-sdk/pull/2276/files#diff-d4bad6c476b1ac95862b5e0a0b1a9195R55-R66).

I'm ok with a deprecation comment, like it is used [here](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.2.0/pkg/runtime/log/log.go#L20).

Comment (joelanford) [reply doc/sdk-cli-reference.md]: I can put the GoDoc deprecation comments on the functions used for `generate openapi` but they won't serve much purpose since none of the functions are public. IMO the big reason to add the deprecation comments is to signal to users that the API is changing, but we're the only users of these private functions.

 This PR isn't really a Go code API deprecation. It's a deprecation of a CLI subcommand.

Comment (camilamacedo86) [reply doc/sdk-cli-reference.md]: Hi @joelanford,
 
The comments would be helpful to us know what should not so longer be used since will be removed. Regards the command, a notice output as it was done [here](https://github.com/kubernetes-sigs/kubebuilder/pull/1137#discussion_r343405975) for example would be more appropriate. 

c/c @estroz

Comment (joelanford) [reply doc/sdk-cli-reference.md]: This PR includes a deprecation notice in the output when `generate openapi` is run. See [here](https://github.com/operator-framework/operator-sdk/pull/2276/files#diff-d4bad6c476b1ac95862b5e0a0b1a9195R55-R66).
